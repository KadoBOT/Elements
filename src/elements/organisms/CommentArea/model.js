export type CommentArea = {
  running: boolean,
  comments: any,
  comment: {
    name: string,
    email: string,
    comment: string,
  },
  errors: {
    name: string,
    email: string,
    comment: string,
  }
}

export type State = CommentArea{}
