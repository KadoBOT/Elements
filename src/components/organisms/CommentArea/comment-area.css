.comment-area {
  margin-top: 70px;
  padding: 0px 70px;
}
