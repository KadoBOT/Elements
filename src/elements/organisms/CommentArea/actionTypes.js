export const requestComments = createAction('request the comments')
export const handleErrors = createAction('handle the errors receveid from the backend') //
export const receiveComments = createAction('receive the comments')
export const postComment = createAction('action to post a new comment')
export const handleForm = createAction('changes the form value on the comment')
export const clearForm = createAction('clear the comment form (default values)')
