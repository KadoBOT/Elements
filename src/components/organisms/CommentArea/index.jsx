import React, {PropTypes} from 'react'
import R from 'ramda'

import { About, Comments, CommentForm } from '../../molecules'
import './comment-area.css'

class CommentArea extends React.Component {
  componentWillMount() {
    const { fetchComments } = this.props
    fetchComments()
  }

  sendComment = () => (
    this.props.sendComment({
      name: this.props.comment.name,
      email: this.props.comment.email,
      comment: this.props.comment.comment
    })
  )

  render() {
    const { comment, comments, errors, running } = this.props
    const isLoading = running || !comments

    if(isLoading) {
      return <div>Loading...</div>
    }

    return(
      <div className="comment-area">
        <About subtitle="Leave a Comment" />
        <CommentForm
          comment={comment}
          errors={errors}
          onChange={this.props.handleForm}
          onSubmit={this.sendComment}
        />
        <div className="comment-area__comments">
          {R.map(c =>
            <Comments key={Math.random()} comment={c} />
          ,comments.results)}
        </div>
      </div>
    )
  }
}

CommentArea.propTypes = {
  fetchComments: PropTypes.func.isRequired,
  comment: PropTypes.object.isRequired,
  comments: PropTypes.any.isRequired,
  errors: PropTypes.object.isRequired,
  running: PropTypes.bool.isRequired,
}

export default CommentArea
