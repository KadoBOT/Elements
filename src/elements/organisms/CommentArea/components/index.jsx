import React, {PropTypes} from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import R from 'ramda'

import { About, Comments, CommentForm } from '../../molecules'
import './comment-area.css'

class CommentArea extends React.Component {
  componentWillMount() {
    const { fetchComments } = this.props
    fetchComments()
  }

  sendComment = () => (
    this.props.sendComment({
      name: this.props.comment.name,
      email: this.props.comment.email,
      comment: this.props.comment.comment
    })
  )

  render() {
    const { comment, comments, errors, running } = this.props
    const isLoading = running || !comments

    if(isLoading) {
      return <div>Loading...</div>
    }

    return(
      <div className="comment-area">
        <About subtitle="Leave a Comment" />
        <CommentForm
          comment={comment}
          errors={errors}
          onChange={this.props.handleForm}
          onSubmit={this.sendComment}
        />
        <div className="comment-area__comments">
          {R.map(c =>
            <Comments key={Math.random()} comment={c} />
          ,comments.results)}
        </div>
      </div>
    )
  }
}

CommentArea.propTypes = {
  fetchComments: PropTypes.func.isRequired,
  comment: PropTypes.object.isRequired,
  comments: PropTypes.any.isRequired,
  errors: PropTypes.object.isRequired,
  running: PropTypes.bool.isRequired,
}

const mapStateToProps = state => ({
  running: state.comments.running,
  comments: state.comments.comments,
  comment: state.comments.comment,
  errors: state.comments.errors,
})

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch)

const CommentAreaContainer = connect(
  createSelector(
    (state, props) => state.comments,
    comments => ({ comments })
  )
)

export default connect(mapStateToProps, mapDispatchToProps)(CommentArea)
