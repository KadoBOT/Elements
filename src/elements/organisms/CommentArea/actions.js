import { createAction } from 'redux-act'
import React from 'react'
import R from 'ramda'

import * as c from './actionTypes'

export const fetchComments = () => dispatch => {
  dispatch(c.requestComments()) // request the comments

  return api.get('comments/').then(R.compose(dispatch, c.receiveComments, R.prop('data'))) //and send another action after its received
}

export const sendComment = payload => dispatch => {
  api.post('comments/', payload) // post a new comment
  .then(() => {
    Toasted( // show the toaster with a success message
      <span>
        Comment Sent Successfully!
      </span>
    )
    dispatch(c.requestComments()) // request the comments again, to update with the created one
    dispatch(c.clearForm()) // clear the inputs
    return api.get('comments/').then(R.compose(dispatch, c.receiveComments, R.prop('data'))) // reiceve the comments including the posted one
  })
  .catch(err => {
    Toasted( // show toaster with error message
      <span>
        Failed to Send Comment... please try again later.
      </span>
    , 'error')
    dispatch(c.handleErrors(err.response.data)) // handle the errors received by the backend
  })
}
