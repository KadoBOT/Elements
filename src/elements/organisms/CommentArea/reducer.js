import { createReducer } from 'redux-act'
import * as c from './actionTypes'
import type { State } from './model'

const comment = {
  name: '',
  email: '',
  comment: ''
}

const errors = {
  name: [],
  email: [],
  comment: []
}

const initialState: State = { running: false, comments: false, comment, errors }

export default createReducer({
  [c.requestComments]: (state) => ({...state, running: true}),
  [c.receiveComments]: (state, comments) => ({ ...state, running: false, comments }),
  [c.handleForm]: (state, evt) => ({
    ...state,
    comment: {
      ...state.comment,
      [evt.target.name]: evt.target.value
    }
  }),
  [c.handleErrors]: (state, errors) => ({ ...state, errors }),
  [c.clearForm]: (state) => ({ ...state, comment, errors })
}, initialState)
