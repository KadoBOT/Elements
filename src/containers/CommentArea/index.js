import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as actions from '../../actions'
import CommentArea from '../../components/organisms/CommentArea'

const mapStateToProps = state => ({
  running: state.comments.running,
  comments: state.comments.comments,
  comment: state.comments.comment
})

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(CommentArea)
