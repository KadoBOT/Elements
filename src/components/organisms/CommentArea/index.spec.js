import React from 'react'
import { shallow, render, mount } from 'enzyme';

import CommentArea from './'

it('should render the CommentArea', () => {
    const wrapper = shallow(
        <CommentArea
          fetchComments={jest.fn()}
          comment={{}}
          comments={false}
          errors={{}}
          running={true}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

it('should render the CommentArea', () => {
    const wrapper = shallow(
        <CommentArea
          fetchComments={jest.fn()}
          comment={{}}
          comments={{
            results: []
          }}
          errors={{}}
          handleForm={jest.fn()}
          running={false}
        />
    );
    expect(wrapper).toMatchSnapshot();
});
