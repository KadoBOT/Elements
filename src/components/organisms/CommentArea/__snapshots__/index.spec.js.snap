exports[`test should render the CommentArea 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    Loading...
</div>,
  "nodes": Array [
    <div>
      Loading...
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <CommentArea
        comment={Object {}}
        comments={false}
        errors={Object {}}
        fetchComments={[Function]}
        running={true} />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": CommentArea {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "comment": Object {},
          "comments": false,
          "errors": Object {},
          "fetchComments": [Function],
          "running": true,
        },
        "refs": Object {},
        "sendComment": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          Loading...
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          Loading...
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CommentArea
    comment={Object {}}
    comments={false}
    errors={Object {}}
    fetchComments={[Function]}
    running={true} />,
}
`;

exports[`test should render the CommentArea 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="comment-area">
    <About
        subtitle="Leave a Comment" />
    <CommentForm
        comment={Object {}}
        errors={Object {}}
        onChange={[Function]}
        onSubmit={[Function]} />
    <div
        className="comment-area__comments">
        
    </div>
</div>,
  "nodes": Array [
    <div
      className="comment-area">
      <About
            subtitle="Leave a Comment" />
      <CommentForm
            comment={Object {}}
            errors={Object {}}
            onChange={[Function]}
            onSubmit={[Function]} />
      <div
            className="comment-area__comments">
            
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <CommentArea
        comment={Object {}}
        comments={
                Object {
                        "results": Array [],
                      }
        }
        errors={Object {}}
        fetchComments={[Function]}
        handleForm={[Function]}
        running={false} />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": CommentArea {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "comment": Object {},
          "comments": Object {
            "results": Array [],
          },
          "errors": Object {},
          "fetchComments": [Function],
          "handleForm": [Function],
          "running": false,
        },
        "refs": Object {},
        "sendComment": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="comment-area">
          <About
                    subtitle="Leave a Comment" />
          <CommentForm
                    comment={Object {}}
                    errors={Object {}}
                    onChange={[Function]}
                    onSubmit={[Function]} />
          <div
                    className="comment-area__comments">
                    
          </div>
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="comment-area">
          <About
                    subtitle="Leave a Comment" />
          <CommentForm
                    comment={Object {}}
                    errors={Object {}}
                    onChange={[Function]}
                    onSubmit={[Function]} />
          <div
                    className="comment-area__comments">
                    
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CommentArea
    comment={Object {}}
    comments={
        Object {
            "results": Array [],
          }
    }
    errors={Object {}}
    fetchComments={[Function]}
    handleForm={[Function]}
    running={false} />,
}
`;
